src/runtime/allocator.ts

```diff
@@ -0,0 +1,291 @@
+import {EntityType, ReferenceType, InterfaceType, SingletonType} from '../types/lib-types.js';
+import {Capabilities} from './capabilities.js';
+import {NewArcOptions, PlanPartition, DeserializeArcOptions} from './arc-info.js';
+import {ArcHost, ArcHostFactory, SingletonArcHostFactory} from './arc-host.js';
+import {RecipeResolver} from './recipe-resolver.js';
+
+export interface Allocator {
+  registerArcHost(factory: ArcHostFactory);
+
+  // TODO(b/182410550): unify `newArc` and `startArc`.
+  // Note: all `newArc` callers will have to support async execution.
+  newArc(options: NewArcOptions): ArcId;
+  startArc(options: NewArcOptions & {planName?: string}): Promise<ArcId>;
+  runPlanInArc(arcId: ArcId, plan: Recipe, reinstantiate?: boolean): Promise<void[]>;
+
+  deserialize(options: DeserializeArcOptions): Promise<ArcId>;
+
+  stopArc(arcId: ArcId);
+
+  // TODO(b/182410550): This method is only called externally when speculating.
+  // It should become private, once Planning is incorporated into Allocator APIs.
+  // Once private, consider not returning a value.
+  assignStorageKeys(arcId: ArcId, plan: Recipe, idGenerator?: IdGenerator): Promise<Recipe>;
+}
+
+export class AllocatorImpl implements Allocator {
+  protected readonly arcHostFactories: ArcHostFactory[] = [];
+  protected readonly arcStateById = new Map<ArcId, {partitions: PlanPartition[], idGenerator: IdGenerator}>();
+  protected readonly hostById: Dictionary<ArcHost> = {};
+
+  constructor(protected readonly runtime: Runtime) {}
+
+  registerArcHost(factory: ArcHostFactory) {
+    this.arcHostFactories.push(factory);
+  }
+
+  newArc(options: NewArcOptions): ArcId {
+    assert(options.arcId || options.arcName);
+    let arcId = null;
+    let idGenerator = null;
+    if (options.arcId) {
+      arcId = options.arcId;
+    } else {
+      idGenerator = IdGenerator.newSession();
+      arcId = idGenerator.newArcId(options.arcName);
+    }
+    assert(arcId);
+    idGenerator = idGenerator || IdGenerator.newSession();
+    if (!this.arcStateById.has(arcId)) {
+      assert(idGenerator, 'or maybe need to create one anyway?');
+      this.arcStateById.set(arcId, {partitions: [], idGenerator});
+    }
+    return arcId;
+  }
+
+  public async startArc(options: NewArcOptions & {planName?: string}): Promise<ArcId> {
+    const arcId = this.newArc(options);
+    await this.runInArc(arcId, options.planName);
+    return arcId;
+  }
+
+  protected async runInArc(arcId: ArcId, planName?: string): Promise<void[]> {
+    assert(this.arcStateById.has(arcId));
+    assert(planName || this.runtime.context.recipes.length === 1);
+    const plan = planName
+      ? this.runtime.context.allRecipes.find(r => r.name === planName)
+      : this.runtime.context.recipes[0];
+    assert(plan);
+    return this.runPlanInArc(arcId, plan);
+  }
+
+  async runPlanInArc(arcId: ArcId, plan: Recipe, reinstantiate?: boolean): Promise<void[]> {
+    plan = await this.resolveRecipe(arcId, plan);
```

Comment:
This should deal with plans, not recipes. OK for that to happen in a future step, just wanted to make sure we're on the same page.

Reply:
I moved most normalization and resolution code into recipe.ts, and some of it back into Runtime (for now).